domain/application/service/metadata.go

```diff
@@ -502,6 +524,66 @@ func verifyRelations(provides, requires, peers map[string]charm.Relation) error
 	return nil
 }
 
+func verifyRelationRoles(provides, requires, peers map[string]charm.Relation) error {
+	for k, v := range provides {
+		if v.Role != charm.RoleProvider {
+			return errors.Errorf("%w; expected provides relation %q to have role %q",
+				applicationerrors.CharmRelationRoleNotValid, k, charm.RoleProvider)
+		}
+	}
+	for k, v := range requires {
+		if v.Role != charm.RoleRequirer {
+			return errors.Errorf("%w; expected requires relation %q to have role %q",
+				applicationerrors.CharmRelationRoleNotValid, k, charm.RoleRequirer)
+		}
+	}
+	for k, v := range peers {
+		if v.Role != charm.RolePeer {
+			return errors.Errorf("%w; expected peers relation %q to have role %q",
+				applicationerrors.CharmRelationRoleNotValid, k, charm.RolePeer)
+		}
+	}
+	return nil
+}
+
+func verifyNoReservedNameMisuses(metadata *internalcharm.Meta, provides, requires, peers map[string]charm.Relation) error {
+	// 'special' charms starting with "juju-" are allowed to use the reserved
+	// interface names.
+	if strings.HasPrefix(metadata.Name, "juju-") {
+		return nil
+	}
+
+	for _, rel := range provides {
+		if rel.Interface == "juju" || strings.HasPrefix(rel.Interface, "juju-") {
+			return errors.Errorf("%w; provides relation %q has reserved interface name %q",
+				applicationerrors.CharmRelationReservedNameMisuse, rel.Name, rel.Interface)
+		}
+	}
+	for _, rel := range peers {
+		if rel.Interface == "juju" || strings.HasPrefix(rel.Interface, "juju-") {
+			return errors.Errorf("%w; peer relation %q has reserved interface name %q",
+				applicationerrors.CharmRelationReservedNameMisuse, rel.Name, rel.Interface)
+		}
+	}
+
+	for name, rel := range requires {
+		// Container-scoped require relations on subordinates are allowed
+		// to use the otherwise-reserved juju-* namespace.
+		if metadata.Subordinate && rel.Scope == charm.ScopeContainer {
+			continue
+		}
+		// TODO(jack-w-shaw): Verify we check the relation name, not interface here.
+		// This has been the case since 2012, but it's unclear why.
+		// https://github.com/juju/charm/commit/f9ad0b589fbface61e45a9687972816c12f92676
```

Comment:
@manadart @SimonRichardson What do you think about this?

Reply:
Chesterton's fence.

Reply:
So, we don't want to prevent the `relation.Interface` value from being prefixed with juju, as that would prevent `juju-info` from being used. I'm unsure why we care about the name being juju. For now, I think we can leave this as is.

I don't think the logic is quite right.

Reply:
Amended